packages/devp2p/src/dpt/kbucket.ts

```diff
@@ -1,56 +1,58 @@
 import { EventEmitter } from 'events'
 import _KBucket = require('k-bucket')
+import { PeerInfo } from './dpt'
 
 const KBUCKET_SIZE = 16
 const KBUCKET_CONCURRENCY = 3
 
-export interface KObj {
-  id?: string
-  port?: string
-  address?: string
+export interface CustomContact extends PeerInfo {
+  id: Uint8Array | Buffer
+  vectorClock: number
 }
 
 export class KBucket extends EventEmitter {
-  _peers: Map<string, any> = new Map()
+  _peers: Map<string, PeerInfo> = new Map()
   _kbucket: _KBucket
-  constructor(id: string | Buffer) {
+  constructor(localNodeId: Buffer) {
     super()
 
-    this._kbucket = new _KBucket({
-      localNodeId: typeof id === 'string' ? Buffer.from(id) : id,
+    this._kbucket = new _KBucket<CustomContact>({
+      localNodeId,
       numberOfNodesPerKBucket: KBUCKET_SIZE,
       numberOfNodesToPing: KBUCKET_CONCURRENCY,
     })
 
-    this._kbucket.on('added', (peer: any) => {
+    this._kbucket.on('added', (peer: PeerInfo) => {
       KBucket.getKeys(peer).forEach((key) => this._peers.set(key, peer))
       this.emit('added', peer)
     })
 
-    this._kbucket.on('removed', (peer: any) => {
+    this._kbucket.on('removed', (peer: PeerInfo) => {
       KBucket.getKeys(peer).forEach((key) => this._peers.delete(key))
       this.emit('removed', peer)
     })
 
-    this._kbucket.on('ping', (...args: any[]) => this.emit('ping', ...args))
+    this._kbucket.on('ping', (oldPeers: PeerInfo[], newPeer: PeerInfo) => {
+      this.emit('ping', { oldPeers, newPeer })
+    })
   }
 
-  static getKeys(obj: Buffer | string | KObj): string[] {
+  static getKeys(obj: Buffer | string | PeerInfo): string[] {
     if (Buffer.isBuffer(obj)) return [obj.toString('hex')]
     if (typeof obj === 'string') return [obj]
 
     const keys = []
     if (Buffer.isBuffer(obj.id)) keys.push(obj.id.toString('hex'))
-    if (obj.address && obj.port) keys.push(`${obj.address}:${obj.port}`)
+    //if (obj.address && obj.port) keys.push(`${obj.address}:${obj.port}`)
```

Comment:
should this remain commented out?

Reply:
Yes, there is just no property `port` on peer objects - just `updPort` and `tcpPort`, so the line in its current form has no effect and I couldn't decide on selecting for one of the other port properties, in fact I am not understanding well enough what this line is doing TBH. So I left to indicate that there was something which might have had some intention. 😄  Could have left a comment though.

Will also leave to not revoke the approval but rather merge here now.